File: crates/deepsec-cli/src/commands/scan.rs
use crate::config_loader::Context;
use crate::file_sources::{FileSourceArgs, resolve};
use anyhow::Result;
use clap::Args as ClapArgs;
use colored::Colorize;
use deepsec_scanner::{ScanOptions, scan, scan_files};

#[derive(Debug, ClapArgs)]
pub struct Args {
    /// Project id (from deepsec.config.toml).
    #[arg(long)]
    pub project_id: String,
    /// Override project root. Defaults to config's project root.
    #[arg(long)]
    pub root: Option<String>,
    /// Only run these matcher slugs (comma-separated).
    #[arg(long)]
    pub matchers: Option<String>,
    /// Skip these matcher slugs (comma-separated).
    #[arg(long)]
    pub skip_matchers: Option<String>,

    /// Explicit file list (repeatable or comma-separated). Mutually
    /// exclusive with --files-from / --diff.
    #[arg(long, value_delimiter = ',')]
    pub files: Option<Vec<String>>,
    /// Read file paths from this file ("-" reads stdin).
    #[arg(long)]
    pub files_from: Option<std::path::PathBuf>,
    /// Scan only files changed vs this git ref (e.g. `origin/main`).
    #[arg(long)]
    pub diff: Option<String>,
}

pub fn run(args: Args, ctx: &Context) -> Result<()> {
    let proj = ctx.project(&args.project_id)?;
    let root = match args.root {
        Some(r) => std::path::PathBuf::from(r),
        None => proj.root.clone(),
    };

    let mut only: Vec<String> = args
        .matchers
        .as_deref()
        .map(super::scan_split_csv)
        .unwrap_or_default();
    let mut exclude: Vec<String> = args
        .skip_matchers
        .as_deref()
        .map(super::scan_split_csv)
        .unwrap_or_default();
    if let Some(cfg) = &ctx.config {
        if only.is_empty() {
            only = cfg.matchers.only.clone();
        }
        if exclude.is_empty() {
            exclude = cfg.matchers.exclude.clone();
        }
    }

    let opts = ScanOptions {
        project_id: args.project_id.clone(),
        root: root.clone(),
        data_root: ctx.data_root.clone(),
        matcher_only: only,
        matcher_exclude: exclude,
        github_url: proj.decl.github_url.clone(),
    };

    let resolved = resolve(
        FileSourceArgs {
            files: args.files.as_deref(),
            files_from: args.files_from.as_deref(),
            diff: args.diff.as_deref(),
        },
        &root,
    )?;

    if let Some(r) = resolved {
        let out = scan_files(&opts, &r.files, &r.source)?;
        println!(
            "{} run={} mode=files source={} files={} candidates={}",
            "scan".bold().green(),
            out.run_id,
            r.source,
            out.files_scanned,
            out.candidate_count
        );
        println!("  tech tags: {}", out.detected.tags.join(", "));
        println!(
            "  matchers active={} skipped={}",
            out.active_matchers.len(),
            out.skipped_matchers.len()
        );
        return Ok(());
    }

    let out = scan(&opts)?;
    println!(
        "{} run={} files={} candidates={}",
        "scan".bold().green(),
        out.run_id,
        out.files_scanned,
        out.candidate_count
    );
    println!("  tech tags: {}", out.detected.tags.join(", "));
    println!(
        "  matchers active={} skipped={}",
        out.active_matchers.len(),
        out.skipped_matchers.len()
    );
    if !out.language_stats.is_empty() {
        println!("  by language:");
        for s in &out.language_stats {
            println!(
                "    {:>12}: {:>5} files  hits in {:>4}",
                s.language, s.files_scanned, s.files_with_match
            );
        }
    }
    Ok(())
}

File: internal/cli/commands/scan.go
package commands

import (
	"fmt"
	"strings"

	"github.com/noeljackson/deepsec/internal/cli"
	"github.com/noeljackson/deepsec/internal/scanner"
	"github.com/spf13/cobra"
)

// NewScanCmd runs the regex scanner.
func NewScanCmd(loader func() (*cli.Context, error)) *cobra.Command {
	var projectID, rootOverride, matchers, skipMatchers, filesFrom, diff string
	var files []string
	cmd := &cobra.Command{
		Use:   "scan",
		Short: "Run the regex scanner over a project",
		RunE: func(cmd *cobra.Command, _ []string) error {
			ctx, err := loader()
			if err != nil {
				return err
			}
			proj, err := ctx.Project(projectID)
			if err != nil {
				return err
			}
			root := proj.Root
			if rootOverride != "" {
				root = rootOverride
			}
			only := splitCSV(matchers)
			exclude := splitCSV(skipMatchers)
			if ctx.Config != nil {
				if len(only) == 0 {
					only = ctx.Config.Matchers.Only
				}
				if len(exclude) == 0 {
					exclude = ctx.Config.Matchers.Exclude
				}
			}

			opts := scanner.Options{
				ProjectID:      projectID,
				Root:           root,
				DataRoot:       ctx.DataRoot,
				MatcherOnly:    only,
				MatcherExclude: exclude,
				GithubURL:      proj.Decl.GithubURL,
			}

			resolved, err := cli.ResolveFiles(cli.FileSourceArgs{
				Files:     files,
				FilesFrom: filesFrom,
				Diff:      diff,
			}, root)
			if err != nil {
				return err
			}
			if resolved != nil {
				out, err := scanner.ScanFiles(opts, resolved.Files, resolved.Source)
				if err != nil {
					return err
				}
				fmt.Printf("scan run=%s mode=files source=%s files=%d candidates=%d\n",
					out.RunID, resolved.Source, out.FilesScanned, out.CandidateCount)
				fmt.Printf("  tech tags: %s\n", strings.Join(out.Detected.Tags, ", "))
				fmt.Printf("  matchers active=%d skipped=%d\n", len(out.ActiveMatchers), len(out.SkippedMatchers))
				return nil
			}

			out, err := scanner.Scan(opts)
			if err != nil {
				return err
			}
			fmt.Printf("scan run=%s files=%d candidates=%d\n",
				out.RunID, out.FilesScanned, out.CandidateCount)
			fmt.Printf("  tech tags: %s\n", strings.Join(out.Detected.Tags, ", "))
			fmt.Printf("  matchers active=%d skipped=%d\n", len(out.ActiveMatchers), len(out.SkippedMatchers))
			if len(out.LanguageStats) > 0 {
				fmt.Println("  by language:")
				for _, s := range out.LanguageStats {
					fmt.Printf("    %12s: %5d files  hits in %4d\n",
						s.Language, s.FilesScanned, s.FilesWithMatch)
				}
			}
			return nil
		},
	}
	cmd.Flags().StringVar(&projectID, "project-id", "", "Project id (required)")
	_ = cmd.MarkFlagRequired("project-id")
	cmd.Flags().StringVar(&rootOverride, "root", "", "Override project root")
	cmd.Flags().StringVar(&matchers, "matchers", "", "Only run these matcher slugs (csv)")
	cmd.Flags().StringVar(&skipMatchers, "skip-matchers", "", "Skip these matcher slugs (csv)")
	cmd.Flags().StringSliceVar(&files, "files", nil, "Explicit file list (csv or repeated)")
	cmd.Flags().StringVar(&filesFrom, "files-from", "", `Read file paths from this file ("-" reads stdin)`)
	cmd.Flags().StringVar(&diff, "diff", "", "Scan only files changed vs this git ref")
	return cmd
}

func splitCSV(s string) []string {
	if s == "" {
		return nil
	}
	parts := strings.Split(s, ",")
	out := make([]string, 0, len(parts))
	for _, p := range parts {
		p = strings.TrimSpace(p)
		if p != "" {
			out = append(out, p)
		}
	}
	return out
}
